File: filter/filter.go
package filter

import (
	"regexp"
	"slices"
	"strings"

	"github.com/pkg/errors"
)

// Validate validates the string filter.
func (f *StringFilter) Validate() error {
	if reSrc := f.GetRe(); reSrc != "" {
		if _, err := regexp.Compile(reSrc); err != nil {
			return errors.Wrap(err, "re")
		}
	}
	return nil
}

// CheckMatch checks if the given value matches the string filter.
func (f *StringFilter) CheckMatch(value string) bool {
	if f == nil {
		return true
	}
	if f.GetEmpty() && value != "" {
		return false
	}
	if f.GetNotEmpty() && value == "" {
		return false
	}
	if val := f.GetValue(); val != "" && value != val {
		return false
	}
	if matchValues := f.GetValues(); len(matchValues) != 0 && !slices.Contains(matchValues, value) {
		return false
	}
	if reSrc := f.GetRe(); reSrc != "" {
		rgx, err := regexp.Compile(reSrc)
		if err != nil {
			// checked in Validate but treat it as a fail
			return false
		}
		if !rgx.MatchString(value) {
			return false
		}
	}
	if prefixSrc := f.GetHasPrefix(); prefixSrc != "" {
		if !strings.HasPrefix(value, prefixSrc) {
			return false
		}
	}
	if suffixSrc := f.GetHasSuffix(); suffixSrc != "" {
		if !strings.HasSuffix(value, suffixSrc) {
			return false
		}
	}
	if containsSrc := f.GetContains(); containsSrc != "" {
		if !strings.Contains(value, containsSrc) {
			return false
		}
	}

	return true
}

File: filter/filter.ts
import { StringFilter } from './filter.pb.js'

/**
 * Validates the string filter.
 * @param filter The string filter to validate
 * @returns Error message if invalid, null if valid
 */
export function validateStringFilter(
  filter: StringFilter | null | undefined,
): string | null {
  if (!filter) {
    return null
  }

  if (filter.re) {
    try {
      new RegExp(filter.re)
    } catch (error) {
      return `re: ${error instanceof Error ? error.message : 'invalid regex'}`
    }
  }

  return null
}

/**
 * Checks if the given value matches the string filter.
 * All of the non-zero rules must match for the filter to match.
 * An empty filter matches any.
 * @param filter The string filter to check against
 * @param value The string value to check
 * @returns True if the value matches the filter, false otherwise
 */
export function checkStringFilterMatch(
  filter: StringFilter | null | undefined,
  value: string,
): boolean {
  if (!filter) {
    return true
  }

  if (filter.empty && value !== '') {
    return false
  }

  if (filter.notEmpty && value === '') {
    return false
  }

  if (filter.value && value !== filter.value) {
    return false
  }

  if (
    filter.values &&
    filter.values.length > 0 &&
    !filter.values.includes(value)
  ) {
    return false
  }

  if (filter.re) {
    try {
      const regex = new RegExp(filter.re)
      if (!regex.test(value)) {
        return false
      }
    } catch {
      // Invalid regex treated as a fail (checked in validate but treat as fail)
      return false
    }
  }

  if (filter.hasPrefix && !value.startsWith(filter.hasPrefix)) {
    return false
  }

  if (filter.hasSuffix && !value.endsWith(filter.hasSuffix)) {
    return false
  }

  if (filter.contains && !value.includes(filter.contains)) {
    return false
  }

  return true
}
